euphony/build.gradle

```diff
@@ -45,38 +46,55 @@ android {
 
     buildFeatures {
         prefab true
+        compose true
+    }
+
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+
+    kotlinOptions {
+        jvmTarget = "1.8"
+    }
+
+    composeOptions {
+        kotlinCompilerExtensionVersion compose_version
     }
 
     testOptions {
         unitTests.returnDefaultValues = true
     }
 
     gradle.buildFinished {
-        def abiArr = ['arm64-v8a', 'armeabi-v7a', 'x86', 'x86_64']
+        def cppBuildDir = ".cxx/Debug"
+        def lastWordIfThereAreNoDevice = "0"
         def logfile = "android_gradle_build_stdout_targets.txt"
-        def lastWordIfThereAreNoDevice = "Linking"
         def resultStr = "\n"
         def noDevice = true
-        for (abi in abiArr) {
-            def filePath = "euphony/.cxx/cmake/debug/$abi/$logfile"
-            def file = new File(filePath)
+
+        fileTree(cppBuildDir).matching {
+            include "**/$logfile"
+        }.forEach( file -> {
             if (file.exists()) {
                 def lines = file.readLines()
-                if (lines.get(lines.size() - 1).contains(lastWordIfThereAreNoDevice))
-                    continue
+                if (lines.get(2).contains(lastWordIfThereAreNoDevice))
+                    return
                 noDevice = false
-                resultStr += "> $abi/$logfile\n"
+                resultStr += "> $file.path\n"
                 resultStr += lines.join("\n")
+                resultStr += "\n\n"
             }
-        }
+        })
+
         if (noDevice)
-            println "The build was successful, not verified by unit tests. For unit testing, connect a virtual device or Android device to adb."
+            println "\nThe build was successful, not verified by unit tests. For unit testing, connect a virtual device or Android device to adb."
         else
-            println "Print build_stdout $resultStr"
+            println "\nPrint build_stdout $resultStr"
     }
```

Comment:
### Here are my last suggestions :)

I added some build error handling code & modify some print messages for unit-test result.
```gradle
gradle.buildFinished {
        def cppBuildDir = ".cxx/Debug"
        def logfile = "android_gradle_build_stdout_targets.txt"
        def resultStr = "\n"
        def isUnitTestWorked = false
        def isDevice = false

        fileTree(cppBuildDir).matching {
            include "**/$logfile"
        }.forEach( file -> {
            if (file.exists()) {
                def lines = file.readLines()
                /* Check the unit-test working */
                if (lines.size() > 0) {
                    def lastSentence = lines.get(lines.size() - 1)
                    if (lastSentence.contains("**** Gtest Success ****")) {
                        isUnitTestWorked = true
                        isDevice = true
                    } else if (lastSentence.contains("ninja: no work to do.")) {
                        isUnitTestWorked = false
                        if (!lines.get(2).contains("0"))
                            isDevice = true
                    } else if (lastSentence.contains("Linking")) {
                        isUnitTestWorked = false
                        return
                    }

                    resultStr += "> $file.path\n"
                    resultStr += lines.join("\n")
                }
            }
        })

        print "\n"
        if (!isDevice) {
            println "THE BUILD WAS SUCCESSFUL, BUT NOT VERIFIED BY NATIVE UNIT TESTS."
            println "FOR UNIT TESTING, PLEASE CONNECT A VIRTUAL DEVICE OR ANDROID DEVICE TO ADB"
        }
        else {
            if (isUnitTestWorked) {
                println "Print build_stdout $resultStr"
                println "\nTHE BUILD & TESTS ARE SUCCESSFUL"
                println "THERE ARE NO ERRORS FOR NATIVE UNIT TESTS."
            }
            else {
                println "THERE ARE NO CHANGES TO THE CODE."
                println "THERE IS NO NEED FOR NATIVE UNIT TESTS."
            }
        }
    }
```

Reply:
Thank you for checking ```build.gradle```.
But, unfortunately, I got an error at below line: :cry: 
```
                        // error message; Index 2 out of bounds for length 2
                        if (!lines.get(2).contains("0"))

```

Reply:
Thanks for sharing printed line! :)

this is real last version I think :)

```groove
gradle.buildFinished {
    def cppBuildDir = ".cxx/Debug"
    def logfile = "android_gradle_build_stdout_targets.txt"
    def resultStr = "\n"
    def isUnitTestWorked = false
    def isDevice = false

    fileTree(cppBuildDir).matching {
        include "**/$logfile"
    }.forEach( file -> {
        if (file.exists()) {
            def lines = file.readLines()
            /* Check the unit-test working */
            if (lines.size() > 0) {
                def lastSentence = lines.get(lines.size() - 1)
                if (lastSentence.contains("**** Gtest Success ****")) {
                    isUnitTestWorked = true
                    isDevice = true
                } else if (lastSentence.contains("ninja: no work to do.")) {
                    isUnitTestWorked = false
                    if (lines.size() > 2 && !lines.get(2).contains("0"))
                        isDevice = true
                } else if (lastSentence.contains("Linking")) {
                    isUnitTestWorked = false
                    return
                }

                resultStr += "> $file.path\n"
                resultStr += lines.join("\n")
            }
        }
    })

    print "\n"
    if (!isDevice) {
        println "THE BUILD WAS SUCCESSFUL, BUT NOT VERIFIED BY NATIVE UNIT TESTS."
        println "FOR UNIT TESTING, PLEASE CONNECT A VIRTUAL DEVICE OR ANDROID DEVICE TO ADB"
    }
    else {
        if (isUnitTestWorked) {
            println "Print build_stdout $resultStr"
            println "\nTHE BUILD & TESTS ARE SUCCESSFUL"
            println "THERE ARE NO ERRORS FOR NATIVE UNIT TESTS."
        }
        else {
            println "THERE ARE NO CHANGES TO THE CODE."
            println "THERE IS NO NEED FOR NATIVE UNIT TESTS."
        }
    }
}
```

Reply:
@designe Thank you! I applied your last version :)